plugins/woocommerce/tests/e2e-pw/global-setup.js

```diff
@@ -50,21 +51,21 @@ module.exports = async ( config ) => {
 	const adminPage = await adminContext.newPage();
 	const customerPage = await customerContext.newPage();
 
-	// Sign in as admin user and save state
+		// Sign in as admin user and save state
 	const adminRetries = 5;
 	for ( let i = 0; i < adminRetries; i++ ) {
 		try {
 			console.log( 'Trying to log-in as admin...' );
 			await adminPage.goto( `/wp-admin` );
 			await adminPage.fill( 'input[name="log"]', admin.username );
 			await adminPage.fill( 'input[name="pwd"]', admin.password );
-			await adminPage.click( 'text=Log In' );
+			await adminPage.click( `text=${getTextForLanguage()['LogIn']}` );
```

Comment:
I may not be following you 100% 😊

The [`getTextForLanguage()`](https://github.com/woocommerce/woocommerce/blob/34899772513ecb18aca9c16d21935fb9ff5c2514/plugins/woocommerce/tests/e2e-pw/test-data/data.js#L543) function has replaced all language-specific references in our scripts. 

It is used in various instances for locators, assertions and some other code instances.

Thanks for the example on overriding the original Playwright functions in JavaScript 👍
Am all ears in on how it could help to reduce maintenance!

Are you suggesting we check all the different instances in which it is used and then override the various usages? 

e.g. as an isolated example, we have code such as the following:
```
await firstVariation
	.getByRole( 'checkbox', {
		name: getTextForLanguage()['Virtual'],
	} )
	.check();
```
Maybe in this case we would override the `getByRole` function to check to see if the passed-in object name (in this case, 'Virtual') has a value that is contained in the language object and if so we replace it with the appropriate value?

We have over 350 uses of the function so we would need to determine all the different ways it was used. some of the many examples are:
page.click 
page.getByLabel
page.check
page.waitForSelector
page.locator
it is also used in various assertions e.g. 
expect( title ).toHaveText( `${getTextForLanguage()['Dashboard']}` );

They are all used in various ways so would take a spot of investigation but is a very interesting idea if it makes things seamless from the point of other e2e test contributions!

Reply:
> Thanks for the example on overriding the original Playwright functions in JavaScript 👍
Am all ears in on how it could help to reduce maintenance!

No problem! This is just a way to reduce the number of changes we'd need to apply in the future. We can follow 2 different approaches (maybe more, but I will focus on these 😆 ):
- We call the getTextForLanguage() function whenever we need to use it, this involves every time we need to traduce every text that we use as a locator, multiple expect methods, etc. This means the getTextForLanguage() will appear hundreds or even thousands of times in our code in the future. If we ever need to refactor something there (add a new attribute, rename it or anything), we'll need a big refactor and a big code review. Renaming refactors is usually easy thanks to IDE features, but still the impact of the change would be high because there would be a lot of occurrences in the code, and there is not 100% guarantee that even a IDE performed rename refactor is always 100% correct, so we would need to review it thoroughly. Or instead...
- We override these built-in Playwright methods to use internally the getTextForLanguage() function and provide a default behaviour if there is no need to use it, this way we'll reduce the number of occurrences in the code. It's like creating a wrapper on top of Playwright but overriding the functions instead to keep using the same methods as always in the code. Basically, instead of writing getTextForLanguage() in every function, just write it on a few of them.

For example, instead of doing something like this (and assuming that Virtual1, Virtual2 and Virtual3 exists and are defined in the language object):
```
await firstVariation
	.getByRole( 'checkbox', {
		name: getTextForLanguage()['Virtual1'],
	} )
	.check();

await secondVariation
	.getByRole( 'checkbox', {
		name: getTextForLanguage()['Virtual2'],
	} )
	.check();

await thirdVariation
	.getByRole( 'checkbox', {
		name: getTextForLanguage()['Virtual3'],
	} )
	.check();
```

We could do something like this:
```
// Override the getByRole method (the code below is kind of pseudocode, I haven't checked how to do it properly)
function getByRoleTranslated(role, options) {
    options.name = getTextForLanguage()['options.name'];
    return getByRole(role, options);
getByRole = getByRoleTranslated;
```

With something like the above done properly, we could just do then:
```
await firstVariation
	.getByRole( 'checkbox', {
		name: 'Virtual1',
	} )
	.check();
	
await secondVariation
	.getByRole( 'checkbox', {
		name: 'Virtual2',
	} )
	.check();
	
await thirdVariation
	.getByRole( 'checkbox', {
		name: 'Virtual3',
	} )
	.check();
```

Thanks to this, we would keep using the native Playwright functions as usual without having to pass new functions as parameters every time we use them, and if we ever need to change anything in the getTextForLanguage() function, like the name of parameters, it's likely that this refactor wouldn't be too painful since we would have fewer occurrences in the code.

By the way, I just learnt about [Playwright Fixtures](https://playwright.dev/docs/test-fixtures), seems it works just like Cypress commands (in case you were familiar with it). Maybe we could use them to implement what I described above. We can even [override existing fixtures](https://playwright.dev/docs/test-fixtures#overriding-fixtures).

I'll ping @lanej0 here in case he could provide more insights about these Playwright fixtures.

Thanks!

Reply:
Great stuff @alopezari - thanks for the clarification 👏🏻😊 - by the end of my ramble I was almost there 😉
I like it - a little bit of pain at the start but long-term gain 👍

Also - thanks for the [Playwright Fixtures](https://playwright.dev/docs/test-fixtures) reference!

I'll check to see all the methods that would be impacted and await the thoughts of @lanej0

Reply:
> by the end of my ramble I was almost there
Yes, you got it! I just realized that my comment wasn't explained very well and decided to develop it a little further 😄 

> One thing to ponder is there are many hard-coded entries across the scripts which have been updated to use interpolation with the function in the middle of the text
Good point! Also the string  that we pass now to the `getTextForLanguage()` function is different than the one we passed to the Playwright methods. For instance, considering your example above, we would have now `LogIn` instead of `Log In`, so we would need to update every single line anyway, which would minimize the benefits of overriding methods.

A potential solution to this in case we decide to override methods might be to keep using the English translation hardcoded in the method and update the getTextForLanguage() function use it in case `LANGUAGE=EN`. If `LANGUAGE!=EN`, then we could find the key of the entry in the `languageObject` object and find the value of the desired translation.

This way we would keep developing tests as usual, without having to look for the name of the translation every time we need to use the getTextForLanguage() function 🤔

Reply:
> A potential solution to this in case we decide to override methods might be to keep using the English translation hardcoded in the method and update the getTextForLanguage() function to directly use it in case LANGUAGE=EN. If LANGUAGE!=EN, then we could look for the key of the entry in the languageObject object by its English translation and find the value of the desired language.
> 
> This way we would keep developing tests as usual, without having to look for the name of the translation every time we need to use the getTextForLanguage() function 🤔

That's a nice idea @alopezari - we'd have to consider longer translations versus subsets e.g. 'Add new' v 'Add new order' and other instances where multiple strings are combined in different orders depending on the langue but this is definitely a good one to think on!

Also - thanks for your great points @rodelgc !

> I would suggest testing our E2E tests against several different languages first to discover quirks from using getTextForLanguage() alone

I think this is a great idea. We currently only have a 'first pass' at one non-English language. It would be worthwhile testing execution against other non-English languages (after determining an efficient way to perform the translation). We also have to consider the quirks of RTL - which will inevitably require more updates. The updates were initially for the execution of the tests of English, Spanish and an RTL such as Arabic but there is definitely a lot of potential here!

> most locators our E2E tests use are already marked for deprecation. For example, page.click(). So we will need to first convert the deprecated methods into the standard ones

great shout out Del - I took a scan through the tests today and can see the `getTextForLanguage()` function is currently used in various ways by the following methods (note: there may be others!):

page.click - deprecated
page.fill - deprecated
locator.filter
page.type - deprecated
page.getByRole
page.getByText
page.dispatchEvent -  deprecated
page.selectOption  deprecated
page.getByPlaceholder
locator.toHaveAttribute



> we will probably need more time and discussion on how both currencies and text translations can be handled inside the overridden methods, considering that getTextForLanguage() only handles texts.

👍

Reply:
I don't have too much to contribute in terms of fixtures in Playwright.  The docs seem to mix them with POM a lot, which I'm not a huge fan of for Playwright -- but they aren't strictly a feature of a POM-based test suite.  

I used them in a limited way when I was working with Cypress, but haven't had the chance to dive into them in too much depth with Playwright.

Reply:
> page.click - deprecated
> page.fill -  deprecated
> locator.filter
> page.type - deprecated
> page.getByRole
> page.getByText
> page.dispatchEvent - deprecated
> page.selectOption deprecated
> page.getByPlaceholder
> locator.toHaveAttribute

I think I also saw something like page.$eval used once or twice, which is deprecated too

Reply:
Good spot Del - I see usages of:
page.$ 
page.$$
page.$$eval
all of which are deprecated as you mention...

Reply:
Another angle to consider regarding putting `getTextForLanguage` inside overridden methods is that we might not be able to completely avoid using CSS class names, HTML id's, or other non-textContent-based strategies in our locators. So if Playwright methods will translate the given text internally, we'll need deeper thought on how `getTextForLanguage` would  interpret the input as whether a text to be translated, or a text that simply refers to a class name or id.